grease/src/Grease/Macaw/ResolveCall.hs

```diff
@@ -228,19 +338,92 @@ lookupFunctionHandle bak la halloc arch memory symMap pltStubs dynFunMap funOvs
+    IO ( C.FnHandle atps (C.StructType rtps)
+       , C.CrucibleState p sym (Symbolic.MacawExt arch) rtp blocks r ctx
+       )
+  )
+
+-- | An implementation of 'LookupSyscallDispatch' that attempts to invoke
+-- syscall overrides if they can be found. If an override cannot be found, the
+-- syscall is simulated as a no-op.
+defaultLookupSyscallDispatch ::
+  OnlineSolverAndBackend solver sym bak t st fs =>
+  bak ->
+  GreaseLogAction ->
+  C.HandleAllocator ->
+  ArchContext arch ->
+  LookupSyscallDispatch p sym arch
+defaultLookupSyscallDispatch bak la halloc arch =
+  LookupSyscallDispatch $ \atps rtps st regs lsr ->
+    case lsr of
+      SkippedSyscall reason -> do
+        -- Treat this syscall as a no-op during simulation.
+        doLog la $ Diag.SkippedSyscall reason
+        let funcName = W4.functionNameFromText "_grease_syscall"
+        handle <- C.mkHandle' halloc funcName atps (C.StructRepr rtps)
+        let override =
+              C.mkOverride'
+                funcName
+                (C.StructRepr rtps)
+                ((arch ^. archSyscallReturnRegisters)
+                   C.UnitRepr (pure ()) atps regs rtps)
+        pure $ useFnHandleAndState handle (C.UseOverride override) st
+      CachedSyscallFnHandle (Stubs.SyscallFnHandle syscallHdl) ->
+        pure (syscallHdl, st)
+      NewSyscall syscallName syscallNum (Stubs.SomeSyscall syscallOv) -> do
+        doLog la $ Diag.SyscallOverride syscallName syscallNum
+        let syscallFnName = W4.functionNameFromText syscallName
+        macawSyscallHdl <- C.mkHandle' halloc syscallFnName atps (C.StructRepr rtps)
+        let macawSyscallOv = macawSyscallOverride bak arch atps rtps syscallOv
+        pure $ useFnHandleAndState macawSyscallHdl (C.UseOverride macawSyscallOv) st
+
+-- | The result of looking up a syscall override.
+data LookupSyscallResult p sym arch atps rtps where
+  -- | The syscall override could not be found for some reason, so it is
+  -- skipped.
+  SkippedSyscall ::
+    -- | The reason for skipping the syscall.
+    SkippedSyscall ->
+    LookupSyscallResult p sym arch atps rtps
+  -- | The syscall has an override that was previously registered.
+  CachedSyscallFnHandle ::
+    -- | The handle for the syscall override.
+    Stubs.SyscallFnHandle '(atps, rtps) ->
+    LookupSyscallResult p sym arch atps rtps
+  -- | The syscall has an override that has not yet been registered.
+  NewSyscall ::
+    -- | The syscall's name.
+    Text ->
+    -- | The syscall's number.
+    Int ->
+    -- | The syscall's override.
+    Stubs.SomeSyscall p sym (Symbolic.MacawExt arch) ->
+    LookupSyscallResult p sym arch atps rtps
+
 -- | An implementation of 'Symbolic.LookupSyscallHandle' that attempts to look
--- up an override for the syscall, and if one is found, invokes it. If one
--- cannot be found, the syscall is simulated as a no-op.
+-- up an syscall override and dispatches on the result.
 lookupSyscallHandle ::
```

Comment:
Similar refactoring suggestion (compute result, call `dispatch` on it) applies here.